Keras callback for pushing models to Hub

merveenoyan opened the pull request:
The callback to push models to Hub. It's a best of both worlds for transformers Keras callback and `push_to_hub_callback()`. I didn't write any tests on it yet because I couldn't decide the best way of doing it. I could either:
1. Test by interfering to training loop to check if the model is pushed in specific intervals.  (which is probably a very very bad idea)
2. Check the commit history.

Here are couple of examples ([push per epoch to a hub model ID](https://huggingface.co/merve/callback-test-2) and [push by URL](https://huggingface.co/merve/callback-push-url/tree/main)). 

I don't want a review yet because 1. it's a branch derived from unapproved model card PR branch (I did it intentionally because I needed model card writing functions inside, so many merge conflicts upcoming lol), which needs merger 2. I didn't write any tests. You're free to try it on your notebooks though.

merveenoyan commented:
I finished the callback, the rest of the mixin file is not up to date (as it is derived from unapproved Keras model card branch) I needed model card writing files. I added one test case that passes at the moment. 
HOWEVER the test is about checking if the files are pushed to Hub or not at the end of the training, not about if it's pushed every step or epoch, which is not a good test.
We can either check the commit history somehow (?) or I will have to somehow interfere to training loop and check if it's being pushed (which is bad). 
I'm asking for an early review. @nateraw @osanseviero

osanseviero commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
+        include_optimizer: Optional[bool] = True,
+        task_name: Optional[str] = None,
+    ):
+        super().__init__()
+        self.save_strategy = save_strategy
+        self.save_steps = save_steps
+        self.repo_path_or_name = repo_path_or_name
+        self.repo_url = repo_url
+        self.organization = organization
+        self.hub_token = hub_token
+        self.api_endpoint = api_endpoint
+        self.model_plot = model_plot
+        self.task_name = task_name
+        self.last_job = None
+        self.git_user = git_user
+        self.git_email = git_email
+        self.include_optimizer = include_optimizer
+        self.task_name = task_name
```
* Please make sure the docstring, the params, and attribute assignment are aligned
* Some of this params are just used in the `__init__`, so it does not give any value to assign them as attributes. Some examples are `repo_url`, `repo_path_or_name`, git user and others

osanseviero commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
+
+        if self.token is None:
+            raise ValueError(
+                "You must login to the Hugging Face hub on this computer by typing `huggingface-cli login` and "
+                "entering your credentials to use `use_auth_token=True`. Alternatively, you can pass your own "
+                "token as the `use_auth_token` argument."
```
This does not hold with my suggestion above, so you need to change it a bit. Also, there is no `use_auth_token` argument

osanseviero commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
+            git_user=self.git_user,
+            git_email=self.git_email,
+        )
+        self.repo.git_pull(rebase=True)
```
Do we need to do git pulling and rebase? `Repository` already clones I think. (there might be a good reason to the rebase which I might be missing though)

osanseviero commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
+                    time.sleep(1)
+            save_pretrained_keras(self.model, self.repo_path_or_name)
+            _create_model_card(
+                self.model, self.repo_path_or_name, self.model_plot, self.task_name
```
Does this require the param to override the model card

nateraw commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -127,15 +134,15 @@ def _create_model_card(
     model_card += "\n</details>"
 
-    if os.path.exists(readme_path):
+    if os.path.exists(readme_path) and from_callback is False:
```
Am I missing something here or are we reading a file at `readme_path` and then writing the same contents right back to `readme_path` again? If that's the case, you can just ignore in this case and just write if the content is different.

If there was a different path you're reading from in order to write, that would make sense (and was sorta what I first expected when I saw this)

merveenoyan replied:
I read it again and thought it was quite funny, I think it was there initially and I didn't want to change the behavior much not to break it that's why I modified it like that lol

nateraw commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
 
+class PushToHubCallback(Callback):
+    """
+    Callback that will save and push Keras models to the Hub regularly. By default, it pushes once per epoch, but this can
```
```suggestion
    Callback that will periodically save and push Keras models to the Hugging Face Hub. By default, it pushes once per epoch, but this can
```

nateraw commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -144,6 +151,172 @@ def _create_model_card(
+    Args:
+        save_strategy (`str`, *optional*, defaults to `"epoch"`):
+            The checkpoint save strategy to adopt during training. Possible values are:
+                - `"no"`: Save is done at the end of training run.
```
Maybe a better strategy name is "end"?

merveenoyan replied:
I named it "end_of_training"

nateraw commented on tests/test_keras_integration.py:
```diff
@@ -180,6 +180,35 @@ def model_fit(self, model):
+        model = self.model_init()
+
+        push_to_hub_callback = PushToHubCallback(
+            save_strategy="epoch",
```
can we make sure we're testing all strategies?

merveenoyan replied:
I will write the tests at the end because I don't know how to test callbacks, I have to check my options and decide. The behavior here we want to test are if they're really saved and pushed in the periods given by user periodically, not only at the end of training so I somehow need to log this info, either through commit history or inside the training.

nateraw commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -38,24 +40,28 @@ def _extract_hyperparameters_from_keras(model):
-        lines.append(values)
+    lines = None
+    if model.history is not None:
+        if model.history.history != {}:
```
Discussion: do we want to print some info to the logger about this case? Not too loudly, ofc, especially if it tends to be a common case. CC @osanseviero

osanseviero replied:
I don't think that's a super valuable addition, but maybe some logging warning is in place for this. 

Btw this seems more related to the parent PR https://github.com/huggingface/huggingface_hub/pull/679/files, it seems the diff is taken an outdated version of that branch.

nateraw reviewed (commented):
Looking good so far. I will give it a try in colab later as well 😄

merveenoyan commented:
I wrote a test that basically at the end of every x, writes last time a commit was made and at the end of training, checks if the frequency matches the length of number of commits.

merveenoyan commented:
I added tests for step strategy and end of training strategies as well.

merveenoyan commented:
I forgot a print statement inside, pls ignore

merveenoyan commented:
I will change the code since currently model card creation is directly called inside `save_pretrained_keras` thought it doesn't make sense to wait until merger, since it will change many things.

merveenoyan commented:
I aligned things with main (that we merged `keras_model_card`) because it might've later affected final tests. Changes include:
> tensorboard override
> model card creation called in `save_pretrained_keras`
I also added override_card arg under `save_pretrained_keras` since card creation was done under it. Did nit picking as well.

merveenoyan commented:
All the tests are failing (except for tensorflow tests) because Callback is not being imported, when I try to import it like so in `keras_mixin.py` 
```python
if is_tf_available():
    import tensorflow as tf
    from tensorflow.keras.callbacks import Callback
```

osanseviero commented:
The issue is that `ValidationCallback` inherits/sub-classes from `Callback`, which is not imported unless you have TensorFlow

osanseviero commented:
Not super elegant, but I think something along these lines would work. There might be a better approach though

```python
if is_tf_available():
    import tensorflow as tf
    callback = tf.keras.callbacks.Callback
else:
    class Dummy(object):
        def __call__(self, *args, **kwargs):
            return None

    callback = Dummy

class ValidationCallback(callback):
```

merveenoyan commented:
@osanseviero I thought problem was something fundamental because while importing tf is no problem importing Callback is a problem so I'll do it for now but keep looking into it

osanseviero reviewed (commented):
The callback looks great to me! I left a minor question, but not much.

I'm a bit unsure if using a `ValidationCallback` is the right approach for testing. Aren't you testing the validation callback but not `PushToHubCallback`? I could easily see them diverging and suddenly you get passing tests for broken functionality.

One option that could work is doing `git log` and see how many results you get from that - that might be a bit more convoluted for testing purposes, but will directly check the results. I'm a bit surprised there are no callback tests in `transformers` :thinking:

merveenoyan commented:
@osanseviero Essentially push to hub callback pushes and validation callback checks if it pushed or not, in sync with push to hub callback's frequencies (and the assertions check these to see if the frequencies are as intended). Maybe my tests are wrong because what I did was that I ran them every time to see the logs and the `lastModified`s of the logs to see if they make sense or not. I wanted to automate this but indeed what you say make sense. I looked to see if I can use any of the attributes of Repository class of Hub and found lastModified so wanted to use it, maybe I was biased.

merveenoyan commented on src/huggingface_hub/keras_mixin.py:
````diff
@@ -144,6 +151,172 @@ def _create_model_card(
+    )
+    model.fit(train_dataset, callbacks=[push_to_hub_callback])
+    ```
+    Args:
````
ah sorry.

merveenoyan commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -135,13 +146,231 @@ def _create_model_card(
+        if self.save_strategy == "epoch":
+            if self.last_job is not None and not self.last_job.is_done:
+                while not self.last_job.is_done:
+                    time.sleep(1)
```
I think I'll remove the batch one since it will be very verbose and add it to on_epoch_end

merveenoyan commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -23,6 +24,15 @@
 if is_tf_available():
     import tensorflow as tf
 
+    callback = tf.keras.callbacks.Callback
+else:
+
+    class Dummy(object):
+        def __call__(self, *args, **kwargs):
+            return None
+
+    callback = Dummy
+
```
I wrote it like this to make the tests pass for the time being.

merveenoyan commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -135,13 +146,231 @@ def _create_model_card(
+        if self.save_strategy == "steps" and batch + 1 % self.save_steps == 0:
+            if self.last_job is not None and not self.last_job.is_done:
+                logger.info("Waiting for existing upload to finish...")
+                while not self.last_job.is_done:
+                    time.sleep(1)
```
So when I didn't make it sleep it was skipping the upload for some reason. The epochs and batches are very fast (which doesn't happen for transformers models but mostly for Keras models)
for "epoch" strategy I was getting logs like:
```
push at epoch 1
push at epoch 4
push at epoch 8
```
`blocking = False` didn't change anything.

merveenoyan commented:
I think I'll just check git log commit messages and make assertion based on it.

merveenoyan commented:
I fixed a bug in steps strategy (fixed the same thing in [transformers](https://github.com/huggingface/transformers/pull/16138)) 
- generate dummy data with np.ones() instead of giving a list
- removed `show_layer_activations` and added tempfile to tests to align with main beforehand
- did nit picking 
I will check once more before I ask for a review.

merveenoyan commented:
The test for steps strategy was flaky for when your number of batches was odd and save steps were even so I rounded to floor since that's what happens with leftover batches.

osanseviero commented on src/huggingface_hub/keras_mixin.py:
```diff
@@ -135,13 +145,181 @@ def _create_model_card(
+                override_card=True,
+                **self.model_save_kwargs,
+            )
+            self.repo.git_add(auto_lfs_track=True)
```
`push_to_hub` already does `git_add` at https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/repository.py#L1324. Same in other places

osanseviero commented on tests/test_keras_integration.py:
```diff
@@ -176,19 +180,149 @@ def test_push_to_hub(self):
+                .split("\n")
+            ),
+            3,
+        )
```
This is much nicer/reliable than having two callbacks!

osanseviero commented on tests/test_keras_integration.py:
```diff
@@ -176,19 +180,149 @@ def test_push_to_hub(self):
+            ),
+            num_epochs + 3,
+        )
+        # epochs, initial commit, end of training commit, one more line
```
Maybe add this comment before the assert. Similarly for the other tests

osanseviero reviewed (commented):
Looks great! I just have some concerns about blocking/waiting before pushing. Left some comments :rocket:

merveenoyan commented:
@osanseviero If model is too big to push then their epochs are long enough to push the model, if it's too small it will be pushed fast as well. (wild guess) Same might not be the case for the `steps` one. We can do what you say for steps and block for epochs.

osanseviero commented:
I usually prefer consistency since it will confuse users that one method always pushes while the other not. Is there any con for making the epoch one not blocking?

merveenoyan commented:
@osanseviero Epochs are taking longer which I think blocking will not cause any change in performance but I'll change now if you think being consistent is better.

osanseviero commented:
Thanks for explaining! Yes, I feel consistency is better, and would do non-blocking which will be appreciated for large models.

merveenoyan commented:
When `blocking` is `False` the user will see:
`
remote: error: cannot lock ref 'refs/heads/main': is at 66d0b2cc64c2c1ad77084cac6c6414fdd2636fc6 but expected adb34ff5f99b94c48df5a435815e2399374bb559 
To https://huggingface.co/merve/callback-batch-test
 ! [remote rejected] main -> main (failed to update ref)
error: failed to push some refs to '....'
`
So HEAD according to the refs is at adb34 and there's a mismatch.
I find it a bit confusing for users. I'll try to fix this through updating ref, I also saw pruning could help.
(I keep this like a diary, you don't have to answer)

osanseviero commented:
I'm a bit surprised for this. Do we get same error in the transformers callback? The code is very similar too.

merveenoyan commented:
@osanseviero I think this is because length of epochs/steps and state of push, it doesn't happen in transformers (because epochs take long), I tested on my own to make the model I'm testing with bigger and it doesn't appear. This also doesn't affect the pushes (I see everything pushed on the other end and the tests pass), it's just confusing for the users.
The models I was testing with previously would take a second or two every step or epoch to train, I think that's why it happened.

osanseviero commented:
Ah I see, this makes sense. I don't think it's a big issue if this would require a significant change in the code to get it working though.

merveenoyan commented:
@nateraw can you review this?

LysandreJik commented:
I think another difference between the implementation between here and `transformers` is that in `transformers` we specifically wait for the previous push to be finished before starting another one. We retrieve a `push_in_progress` [variable](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2825-L2827) that we [check](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2796-L2797) before trying to push again. Could that be the source of the refs mismatch?

merveenoyan commented:
@LysandreJik yes, I wanted it to push every commit, it makes sense to disable that as well. (then my tests will be a bit weird given I don't push every commit, but it should be fine)

merveenoyan commented:
@LysandreJik I get the same thing even when I return the `last_job` from the push.

LysandreJik commented:
Removing this PR from the v0.5 milestone as version v0.5 will be released in a bit.

merveenoyan commented:
I will come up with a way to test this without messing refs and making model bigger.

BenjaminBossan commented:
Let me preface this by admitting that I have very little experience with keras but here are my thoughts:

As a user, when I would see a callback like this, I would expect it to work similarly to [ModelCheckpoint](https://keras.io/api/callbacks/model_checkpoint/). E.g. there, I have the option to monitor a specific metric like validation loss and only create a checkpoint if that metric improves.

Implementation-wise, I took a look at [`ModelCheckpoint` code](https://github.com/keras-team/keras/blob/07e13740fd181fc3ddec7d9a594d8a08666645f6/keras/callbacks.py#L1162) to see if it could be used as a base class with a few changes, but it does not encapsulate the i/o part, so this would probably be a bad idea. Still, even if a lot of code would need to be copy-pasted, at least we can have confidence that `ModelCheckpoint` has a long history of usage and is well tested, allowing `PushToHubCallback` to focus on the i/o.

nateraw commented:
@BenjaminBossan thanks for the feedback! This is a really great idea, and I ran into it too when I made a callback for PyTorch Lightning. Almost better off copying the default model checkpoint callback and adding the few lines for pushing somewhere within. would love to hear others thoughts

BenjaminBossan commented:
@merveenoyan I added a feature ([this PR](https://github.com/skorch-dev/skorch/pull/881)) on skorch to save model checkpoints on HF Hub. I took a different approach there which works with _existing callbacks_ instead of writing a new one. Maybe that could be a more elegant solution for keras as well? Take a look at [this notebook](https://github.com/skorch-dev/skorch/blob/feature/store-models-on-HF-hub-and-docs/notebooks/Hugging_Face_Model_Checkpoint.ipynb) to see how it works from a user perspective. Whether such an approach would work with keras `ModelCheckpoint`, I can't say for sure.

merveenoyan commented:
[This PR](https://github.com/skorch-dev/skorch/pull/881) by @BenjaminBossan uses `upload_file` which makes sense to use for this case, then it doesn't mess up git ref (from what I see and my guess) Should I wait until we migrate mixins to that or would you like me to do it soon? @osanseviero @LysandreJik @nateraw

osanseviero commented:
I would wait until #847 is merged, @Wauplin is actively working on this

Wauplin commented:
@merveenoyan I don't know what "soon" means and how urgent is this callback feature, but just an update to say that keras mixins to leverage non-git uploads are on their way (see PR https://github.com/huggingface/huggingface_hub/pull/847). Especially it will be easier to build a callback based on `push_to_hub_keras` since there will be no need for an initialization with a `git pull` anymore. See also API example from @LysandreJik in https://github.com/huggingface/huggingface_hub/pull/847#issue-1210196213 .

osanseviero commented:
Yes, I don't think this has been widely requested or urgent, and as waiting will make everyone's life easier I would just wait.

Wauplin commented:
Got back to this PR while cleaning issues on the repo. Too bad we never got back to it :confused: 
Since the Keras integration is directly in Keras now (huggingface_hub's integration being only available with TF-Keras/Keras2), I suggest to close this PR
